homeassistant/components/number/__init__.py

```diff
@@ -99,9 +99,30 @@ async def async_set_value(entity: NumberEntity, service_call: ServiceCall) -> No
     """Service call wrapper to set a new value."""
     value = service_call.data["value"]
     if value < entity.min_value or value > entity.max_value:
-        raise ValueError(
-            f"Value {value} for {entity.entity_id} is outside valid range"
-            f" {entity.min_value} - {entity.max_value}"
+        raise ServiceValidationError(
+            translation_domain=DOMAIN,
+            translation_key="out_of_range",
+            translation_placeholders={
+                "value": value,
+                "entity_id": entity.entity_id,
+                "min_value": str(entity.min_value),
+                "max_value": str(entity.max_value),
+            },
+        )
+
+    # With a min step size of 0.001, the modulus of (value - offset) * 1000
+    # and the stepsize * 100 should never be greater than 1
+    if round((value - entity.min_value) * 1000 % (entity.step * 1000), 3) >= 1:
```

Comment:
I don't understand the comment and I don't understand the code. Why do we multiply with 1000, that seems completely arbitrary? Why isn't the allowed error scaled with the step size?